lib/softoken/kbkdf.c

```diff
@@ -0,0 +1,849 @@
+            } else {
+                /* It is safe to pass chaining_ptr when the KDF is in counter
+                 * mode. */
+                PR_ASSERT(chaining_prf != NULL);
+                ret = sftk_MAC_Update(ctx, chaining_prf, ctx->output_size);
+            }
+            break;
+        case CK_SP800_108_COUNTER:
+            /* Only present in the case when not using COUNTER mode. */
+            PR_ASSERT(mech != CKM_SP800_108_COUNTER_KDF);
+            break;
+        case CK_SP800_108_BYTE_ARRAY:
+            ret = sftk_MAC_Update(ctx, (CK_BYTE_PTR) param->pValue, param->ulValueLen);
+            break;
+        case CK_SP800_108_DKM_LENGTH: {
+            /* We've already done the hard work of calculating the length in
+             * the kbkdf_CalculateIterations function; we merely need to add
+             * the length to the desired point in the input stream. */
+            CK_SP800_108_DKM_LENGTH_FORMAT_PTR length_format = (CK_SP800_108_DKM_LENGTH_FORMAT_PTR) param->pValue;
+            CK_BYTE buffer[8];
+            CK_ULONG num_bytes;
+            kbkdf_EncodeInteger(length, length_format->ulWidthInBits, length_format->bLittleEndian, buffer, &num_bytes);
+            ret = sftk_MAC_Update(ctx, buffer, num_bytes);
+            break;
+                                      }
+        default:
+            PR_ASSERT(PR_FALSE);
+            break;
+        }
+
+        if (ret != CKR_OK) {
+            PR_ASSERT(PR_FALSE);
+            return ret;
+        }
+    }
+
+    return CKR_OK;
+}
+
+CK_RV kbkdf_SaveKey(SFTKObject *key, uint8_t *key_buffer, size_t key_len) {
+    /* This is at least necessary under kbkdf_CreateKey; might as well always
+     * do it. TODO: Figure out if there's any harm in always doing it. */
+    CK_OBJECT_CLASS classType = CKO_SECRET_KEY;
+    /* Safe to ignore ret? */
+    sftk_forceAttribute(key, CKA_CLASS, &classType, sizeof(classType));
+    return sftk_forceAttribute(key, CKA_VALUE, key_buffer, key_len);
+}
+
+CK_ULONG kbkdf_GetDerivedKeySize(CK_DERIVED_KEY_PTR derived_key) {
+    /* Precondition: kbkdf_ValidateDerived(...) returns CKR_OK for this key,
+     * which implies that keySize is defined. */
+
+    CK_KEY_TYPE keyType = CKK_GENERIC_SECRET;
+    CK_ULONG keySize = 0;
+
+    for (size_t offset = 0; offset < derived_key->ulAttributeCount; offset++) {
+        CK_ATTRIBUTE_PTR template = derived_key->pTemplate + offset;
+
+        /* Find the two attributes we care about. */
+        if (template->type == CKA_KEY_TYPE) {
+            keyType = *(CK_KEY_TYPE *)template->pValue;
+        } else if (template->type == CKA_VALUE_LEN) {
+            keySize = *(CK_ULONG *)template->pValue;
+        }
+    }
+
+    /* Prefer keySize, if we have it. */
+    if (keySize > 0) {
+        return keySize;
+    }
+
+    /* Else, fall back to this mapping. We know kbkdf_ValidateDerived(...)
+     * passed, so this should return non-zero. */
```

Comment:
If keyType is CKK_GENERIC_SECRET, then sftk_MapKeySize may return '0['. Is there a default, (like the size of the underlying prf)? or does the mechanism require the length and fails with invalid template if it doesn't exist?

Reply:
In `NSC_DeriveKey`, we get passed a `(key, keySize)` tuple that has validated `keySize != 0` if there is a default:

https://github.com/nss-dev/nss/blob/master/lib/softoken/pkcs11c.c#L6767-L6769

---

In `kbkdf_ValidateParameters(...)`, we validate the same against each derived key.